x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java

```diff
@@ -0,0 +1,145 @@
+    @UpdateForV9 // use MINUS_ONE once that means no timeout
+    private static final TimeValue MASTER_TIMEOUT = TimeValue.MAX_VALUE;
+    private volatile boolean licenseStateListenerRegistered;
+
+    public void init() {
+        listenForLicenseStateChanges();
+        clusterService.addListener(this);
+    }
+
+    void listenForLicenseStateChanges() {
+        assert licenseStateListenerRegistered == false : "listenForLicenseStateChanges() should only be called once";
+        licenseStateListenerRegistered = true;
+        licenseState.addListener(this);
+    }
+
+    @Override
+    public void licenseStateChanged() {
+        licenseIsValid = ENTERPRISE_GEOIP_FEATURE.checkWithoutTracking(licenseState);
+        maybeUpdateTaskState(clusterService.state());
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        hasIngestGeoIpMetadata = event.state().metadata().custom(INGEST_GEOIP_CUSTOM_METADATA_TYPE) != null;
+        final boolean ingestGeoIpCustomMetaChangedInEvent = event.metadataChanged()
+            && event.changedCustomMetadataSet().contains(INGEST_GEOIP_CUSTOM_METADATA_TYPE);
+        final boolean masterNodeChanged = Objects.equals(
+            event.state().nodes().getMasterNode(),
+            event.previousState().nodes().getMasterNode()
+        ) == false;
+        /*
+         * We don't want to potentially start the task on every cluster state change, so only maybeUpdateTaskState if this cluster change
+         * event involved the modification of custom geoip metadata OR a master node change
+         */
+        if (ingestGeoIpCustomMetaChangedInEvent || (masterNodeChanged && hasIngestGeoIpMetadata)) {
+            maybeUpdateTaskState(event.state());
+        }
+    }
+
+    private void maybeUpdateTaskState(ClusterState state) {
+        // We should only start/stop task from single node, master is the best as it will go through it anyway
+        if (state.nodes().isLocalNodeElectedMaster()) {
+            if (licenseIsValid) {
+                if (hasIngestGeoIpMetadata) {
+                    ensureTaskStarted();
+                }
+            } else {
+                ensureTaskStopped();
+            }
+        }
+    }
+
+    private void ensureTaskStarted() {
+        assert licenseIsValid : "Task should never be started without valid license";
+        persistentTasksService.sendStartRequest(
+            ENTERPRISE_GEOIP_DOWNLOADER,
+            ENTERPRISE_GEOIP_DOWNLOADER,
+            new EnterpriseGeoIpTaskParams(),
+            MASTER_TIMEOUT,
+            ActionListener.wrap(r -> logger.debug("Started enterprise geoip downloader task"), e -> {
+                Throwable t = e instanceof RemoteTransportException ? ExceptionsHelper.unwrapCause(e) : e;
+                if (t instanceof ResourceAlreadyExistsException == false) {
+                    logger.error("failed to create enterprise geoip downloader task", e);
+                }
+            })
+        );
```

Comment:
We expect geoip metadata changes to be really rare (there are only 7 possible databases to download right now). And the task start request appears to be very lightweight. So it seemed a fair tradeoff, rather than a request to check if the task was running. Do you know of an inexpensive way to check if a task is running?

Reply:
I talked through this one with @original-brownbear and he had some ideas on how we could write things to be better -- but he said this was fine enough as it was and that the better version (while better) is not urgently required to replace what's here.